### PR TITLE
std: Start implementing wasm32 atomics

### PR DESCRIPTION
**Pull request by alexcrichton**
This commit is an initial start at implementing the standard library for
wasm32-unknown-unknown with the experimental `atomics` feature enabled. None of
these changes will be visible to users of the wasm32-unknown-unknown target
because they all require recompiling the standard library. The hope with this is
that we can get this support into the standard library and start iterating on it
in-tree to enable experimentation.

Currently there's a few components in this PR:

* Atomic fences are disabled on wasm as there's no corresponding atomic op and
  it's not clear yet what the convention should be, but this will change in the
  future!
* Implementations of `Mutex`, `Condvar`, and `RwLock` were all added based on
  the atomic intrinsics that wasm has.
* The `ReentrantMutex` and thread-local-storage implementations panic currently
  as there's no great way to get a handle on the current thread's "id" yet.

Right now the wasm32 target with atomics is unfortunately pretty unusable,
requiring a lot of manual things here and there to actually get it operational.
This will likely continue to evolve as the story for atomics and wasm unfolds,
but we also need more LLVM support for some operations like custom `global`
directives for this to work best.

### CODE REVIEW

**Comment by TimNN** [comment]
Ping from triage @shepmaster: This PR requires your review.

File: src/libcore/sync/atomic.rs [hunk 1]
```diff
@@ -2231,7 +2231,15 @@ unsafe fn atomic_umin<T>(dst: *mut T, val: T, order: Ordering) -> T {
 /// [`Relaxed`]: enum.Ordering.html#variant.Relaxed
 #[inline]
 #[stable(feature = "rust1", since = "1.0.0")]
+#[cfg_attr(target_arch = "wasm32", allow(unused_variables))]
 pub fn fence(order: Ordering) {
+    // On wasm32 it looks like fences aren't implemented in LLVM yet in that
```
Comment (mzji) [inline src/libcore/sync/atomic.rs]: Should we add a `FIXME` here?

**Comment by alexcrichton** [comment]
ping r? @sfackler

File: src/libstd/sys/wasm/condvar_atomics.rs [hunk 2]
```diff
@@ -0,0 +1,103 @@
...
+
+    #[inline]
+    pub unsafe fn destroy(&self) {
+        // nothing to do
+    }
+
+    #[inline]
+    fn ptr(&self) -> *mut i32 {
+        assert_eq!(mem::size_of::<usize>(), mem::size_of::<i32>());
+        &self.cnt as *const AtomicUsize as *mut usize as *mut i32
```
Comment (sfackler) [inline src/libstd/sys/wasm/condvar_atomics.rs]: It seems like there's one more cast than necessary here, right?

File: src/libstd/sys/wasm/rwlock_atomics.rs [hunk 3]
```diff
@@ -0,0 +1,162 @@
...
+            State::Writing => invalid(),
+        };
+        if zero {
+            *self = State::Unlocked;
+        }
+        zero
+    }
+
+    fn dec_writers(&mut self) {
+        let zero = match *self {
```
Comment (sfackler) [inline src/libstd/sys/wasm/rwlock_atomics.rs]: `zero` is () here.

File: src/libstd/sys/wasm/rwlock_atomics.rs [hunk 4]
```diff
@@ -0,0 +1,162 @@
...
+unsafe impl Send for RWLock {}
+unsafe impl Sync for RWLock {}
+
+// This rwlock implementation is a relatively simple implementation which has a
+// condition variable for readers/writers as well as a mutex protecting the
+// internal state of the lock. A current downside of the implementation is that
+// unlocking the lock will notify *all* waiters rather than just readers or just
+// writers. This can cause lots of "thundering stampede" problems. While
+// hopefully correct this implementation is very likely to want to be changed in
+// the future.
```
Comment (sfackler) [inline src/libstd/sys/wasm/rwlock_atomics.rs]: This implementation additionally doesn't provide any fairness guarantees, right?

File: src/libstd/sys/wasm/condvar_atomics.rs [hunk 5]
```diff
@@ -0,0 +1,103 @@
...
+        // 0 == woken, 1 == not equal to `ticket`, 2 == timeout (shouldn't happen)
+        debug_assert!(val == 0 || val == 1);
+        mutex.lock();
+    }
+
+    pub unsafe fn wait_timeout(&self, mutex: &Mutex, dur: Duration) -> bool {
+        let ticket = self.cnt.load(SeqCst) as i32;
+        mutex.unlock();
+        let nanos = dur.as_nanos();
+        assert!(nanos <= i64::max_value() as u128);
```
Comment (sfackler) [inline src/libstd/sys/wasm/condvar_atomics.rs]: We assert here but truncate below in sleep on duration overflow. It seems like truncation is probably more reasonable behavior?

File: src/libcore/sync/atomic.rs (hunk 1 above)
Comment (alexcrichton) [reply src/libcore/sync/atomic.rs]: @sfackler unfortunately this is used by `Arc` which means that apps quickly stop working if it fails to compile or panics :(

Alternatively we could remove libstd's usage of fences on wasm, but this seemed like a smaller local change for now

**Pull request merged**